File: app/src/main/java/com/arthur/gloria/logired/core/websocket/WebSocketManager.kt
package com.arthur.gloria.logired.core.websocket

import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

class WebSocketManager {

    private var webSocket: WebSocket? = null

    private val client: OkHttpClient by lazy {
        val trustAllCerts = arrayOf<TrustManager>(object : X509TrustManager {
            override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {}
            override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {}
            override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
        })
        val sslContext = SSLContext.getInstance("SSL").apply {
            init(null, trustAllCerts, java.security.SecureRandom())
        }
        OkHttpClient.Builder()
            .sslSocketFactory(sslContext.socketFactory, trustAllCerts[0] as X509TrustManager)
            .hostnameVerifier { _, _ -> true }
            .readTimeout(0, TimeUnit.MILLISECONDS)
            .pingInterval(20, TimeUnit.SECONDS)
            .build()
    }

    var onOpen: (() -> Unit)? = null
    var onMessage: ((String) -> Unit)? = null
    var onFailure: ((String) -> Unit)? = null
    var onClosed: (() -> Unit)? = null

    fun connect(url: String) {
        disconnect()
        val request = Request.Builder().url(url).build()
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                onOpen?.invoke()
            }
            override fun onMessage(webSocket: WebSocket, text: String) {
                onMessage?.invoke(text)
            }
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                onFailure?.invoke(t.message ?: "Error WebSocket")
            }
            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                onClosed?.invoke()
            }
        })
    }

    fun send(message: String): Boolean = webSocket?.send(message) ?: false

    fun disconnect() {
        webSocket?.close(1000, "Closing")
        webSocket = null
    }
}

File: lib/core/websocket/websocket_manager.dart
import 'dart:async';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/io.dart';

class WebSocketManager {
  WebSocketChannel? _channel;
  String? _url;
  Timer? _retryTimer;
  bool _closed = false;

  void Function()? onOpen;
  void Function(String)? onMessage;
  void Function(String)? onFailure;
  void Function()? onClosed;

  static const _retryDelay = Duration(seconds: 4);

  void connect(String url) {
    _url = url;
    _closed = false;
    _open();
  }

  void _open() {
    _retryTimer?.cancel();
    _channel?.sink.close(1000);
    _channel = null;
    if (_closed || _url == null) return;
    try {
      _channel = IOWebSocketChannel.connect(
        Uri.parse(_url!),
        pingInterval: const Duration(seconds: 20),
      );
      onOpen?.call();
      _channel!.stream.listen(
        (message) => onMessage?.call(message.toString()),
        onError: (error) {
          onFailure?.call(error.toString());
          _scheduleRetry();
        },
        onDone: () {
          onClosed?.call();
          _scheduleRetry();
        },
      );
    } catch (e) {
      onFailure?.call(e.toString());
      _scheduleRetry();
    }
  }

  void _scheduleRetry() {
    if (_closed) return;
    _retryTimer?.cancel();
    _retryTimer = Timer(_retryDelay, _open);
  }

  void send(String message) {
    _channel?.sink.add(message);
  }

  void disconnect() {
    _closed = true;
    _retryTimer?.cancel();
    _retryTimer = null;
    _channel?.sink.close(1000);
    _channel = null;
  }
}
